File: problem-set-6/sentimental-readability/readability.py
from cs50 import get_string;

def main():
    # Prompt the user to enter a text string
    text = get_string("Text: ")

    # Count the total number of letters in the input text
    letters = count_letters(text)
    # Count the total number of words in the input text
    words = count_words(text)
    # Count the total number of sentences in the input text
    sentences = count_sentences(text)

    # Calculate the average number of letters per 100 words
    L = (letters / words) * 100
    # Calculate the average number of sentences per 100 words
    S = (sentences / words) * 100

    # Compute the Coleman-Liau index using the formula:
    # index = 0.0588 * L - 0.296 * S - 15.8
    index = 0.0588 * L - 0.296 * S - 15.8
    # Round the index to the nearest whole number to get the grade level
    grade = round(index)

    # Determine the readability grade level to print
    if grade < 1:
        # If grade is less than 1, print special message
        print("Before Grade 1")
    elif grade >= 16:
        # If grade is 16 or higher, print "Grade 16+"
        print("Grade 16+")
    else:
        # Otherwise, print the grade number
        print(f"Grade {grade}")


def count_letters(text):
    count = 0
    # Iterate over every character in the text
    for char in text:
        # Check if the character is an alphabetical letter (A-Z or a-z)
        if char.isalpha():
            count += 1  # Increment letter count if true
    return count  # Return total letters counted


def count_words(text):
    # Split the text into words by spaces (assuming words are separated by spaces)
    words = text.split()
    # Return the total number of words
    return len(words)


def count_sentences(text):
    count = 0
    # Iterate over every character in the text
    for char in text:
        # Check if the character indicates end of a sentence
        if char in ['.', '!', '?']:
            count += 1  # Increment sentence count
    return count  # Return total sentences counted


if __name__ == "__main__":
    main()

File: problem-set-2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Get the text
    string text = get_string("Text: ");

    // Counting ingredients
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Index calculations
    float L = (float) letters / words * 100;
    float S = (float) sentences / words * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;

    // Rounding up
    int grade = round(index);

    // Writing out the result
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    };
};

// Counts letters (a-z, A-Z)
int count_letters(string text)
{
    int count = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isalpha(text[i]))
        {
            count++;
        }
    }
    return count;
};

// Word counts (separated by spaces)
int count_words(string text)
{
    int count = 1; // The first word begins without spaces
    for (int i = 0; i < strlen(text); i++)
    {
        if (isspace(text[i]))
        {
            count++;
        }
    }
    return count;
};

// Count sentences (ending with ., ! or ?)
int count_sentences(string text)
{
    int count = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count++;
        }
    }
    return count;
};
